proposals/0047-last-hard-fork-syscall.md

```diff
@@ -0,0 +1,110 @@
+participating in the restart effort. Currently can consider the slot on which
+hardfork was done as a restart slot.
+
+Program developers may find it useful to know that a hard fork has recently
+occurred. This information can help them prevent arbitrage and liquidation
+caused by outdated oracle price account states. However, the cluster's restart
+process takes several hours; in that time, the world can change, causing asset
+prices to fluctuate. As the cluster updates the state of all accounts, malicious
+actors could take advantage of the delay by conducting trades using outdated
+prices, anticipating that they will soon catch up to the actual prices of
+real-world assets. Knowing that cluster restart has been done recently, programs
+can manage these cases more appropriately.
+
+## Alternatives Considered
+
+No alternate considerations; we need to have the value of last restart slot
+while executing the program to correctly treat this case. We cannot have an
+account because then it should be updated just after the restart is successful,
+which will add complexity. The best way is to create a new syscall to get this
+information during execution of transaction, which will help us get last
+restart slot without any interface changes for the instruction.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+Currently in Solana Labs validator client `hard fork` slots are a good indicator
+when cluster was restarted. This may change in the future following criteas
+should be met to choose restart slot.
+
+* Should be indicative of when cluster was restarted
+* Should be monotonically increasing
+* Sould be less than equal to current slot
+
+The next part will consider hardfork slot is equal to restart slot.
+
+### Creation of a new syscall
+
+The implementation of this syscall is pretty straitforward. In Solana Labs
+validator client all the hardforks for a cluster are stored in the bank
+structure. The last hard fork slot can be retrieved and then stored in invoke
+context, so that the executing program can access it.
+
+For other validator clients, we have to get the last hard fork slot information
+and make it accessible to the runtime of the program. If there is no hard fork
+done yet on the cluster we consider that the first hard for is at Slot `0`.
+
+### Overview of changes for solana client
+
+The hardfork data is available in the `Bank`. The structure `HardForks` contains
+a vector with all the previous hard forks. The vector is also sorted when we
+register a slot. So the last element of the vector is usually the last slot at
+which hard fork was done. We can get the last hard fork slot from the bank and
+pass it to invoke context structure. We can register new syscall in the file
+`definitions.rs` where other syscalls are defined.
+
+```rust
+define_syscall!(fn sol_get_last_restart_slot() -> Slot);
+```
+
+We can then return the data of last hardfork slot passed to the invoke context
+in this functions implementation.
```

Comment:
I think the signature of the new syscall should match between clients.

Reply:
I am not sure what to add as test vectors.

Reply:
@mvines @t-nelson what do you guys think?

I have provided syscall name, I can add syscall id (murmur3 hash). 

I am correct we should not any code or pseudo code to the SIMD.

Reply:
so this implies the implementation section appears to be repetitive?? The SIMD should clearly outline the runtime modifications and provide some sample test vectors. However, it is at the discretion of each client to determine how to incorporate them

Reply:
Ok understood thanks for the details.

Reply:
> @godmodegalactus @gauravjaincr7 I guess pseudocode is optional here since the algorithm is trivial (it just returns some externally-set value). My original point was that SIMDs should be implementation-agnostic (i.e. not refer to any Rust source files of a specific client). Because otherwise this SIMD only makes sense to devs who are familiar with that particular implementation. Things like `definitions.rs` might also be renamed in the future. Instead, I suggest describing the actual changes to the protocol.
> 
> Which seems to be:
> 
>     1. Feature gate: Since this change is a hardfork itself, protect it with a feature gate
> 
>     2. eBPF instruction set (verifier now accepts one more immediate value for the call instruction, i.e. the syscall ID); New builtin syscall handler and compute unit cost component
> 
>     3. Snapshot format: Changing the bank data structure might introduce an additional field to the snapshot protobuf, but I'm not sure if that's the case here
> 
>     4. Bank hash: If `solana-validator` is restarting from a snapshot from social consensus, the bank hash needs to change to indicate a hardfork
> 
> 
> Something like
> 
> ```
> ### Feature Gate
> 
> The runtime changes made by this proposal are activated by a feature gate with the address:
> 
>     <ADDRESS>
> 
> Upon activation, the functionality described below immediately becomes available.
> 
> ### New syscall
> 
> Newly deployed and upgraded programs may invoke syscall `sol_get_last_restart_slot` (syscall ID <0x...>).
> Each invocation consumes the syscall base amount of compute units (`<number>` on Solana mainnet).
> 
> Register `r0` is set to the last restart slot number (an unsigned 64-bit integer). All other GPRs are not modified.
> 
> Pseudocode:
> 
> # consumes given amount of compute units and may abort execution if compute units are exceeded
> fn consume_cpu(amount)
> 
> # Syscall handler
> fn sol_get_last_restart_slot():
>     consume_cu(SYSCALL_BASE_AMOUNT)
>     r0 = last_restart_slot
> 
> The syscall will be available in architectures eBPF and SBF. (TODO: Ask @Lichtso if this should also be available in SBFv2)
> ```

Thanks for details. By the way it raises a question, do we really need to have a feature gate for this feature?

Reply:
Yeah I understand, yes this is highly likely if this feature is not added without any feature gate.

Ok noted I will add it too.

Reply:
Added some more information as requested.